### ISSUE TITLE
[BUG]Restore will get stuck if the restore error is not captured

### ISSUE THREAD

**Issue by shuo-wu** [longhorn/longhorn#1188]
**Describe the bug**
Right now some errors caused by replica restore are not captured by the longhorn manager and not recorded in engine restore status. Then there is no way to handle the error and continue the restore.

**To Reproduce**
1. Set backupstore for Longhorn
2. Create a regular Longhorn volume with the workload, then write data to the volume.
3. Create a backup for the volume, then launch a DR volume based on the backup.
4. Wait for the DR volume init restore complete. Then create a non-empty directory named `volume-delta-<last restored backup name>.img` for one replica of the DR volume. This directory will fail the following incremental restoration for the replica. (e.g., `mkdir -p <replica data path>/volume-delta-backup-c3edca9817c14f1a.img/dir`)
5. Write data to the regular volume then create a backup. Then wait for the DR volume incremental restoration result. 
6. Check `restoreStatus` and `lastRestoredBackup` in the DR volume engine status. There is no   `restoreStatus` and `engine.status.lastRestoredBackup` is not updated.
7. Create 2 more backups for the regular volume. The DR volume won't continue to restore volume.

**Expected behavior**
1. The error should be recorded in `restoreStatus` and `engine.status.lastRestoredBackup` should be updated.
2. The replica that fails to restore the data will be marked as failed

**Environment:**
 - Longhorn version: v1.0.0
 - Kubernetes version: 
 - Node OS type and version:

**Comment by shuo-wu** [comment]
Integration test plan:
Test if the replica fails the incremental restore will be marked as ERROR.
1. Set a random the backupstore.
2. Create a regular volume and a backup (with some data).
3. Create a DR volume from the backup.
4. Create the 2nd backup for the regular volume
5. Wait for the DR volume incremental restore complete.
6. Pick up a replica of the DR volume, Create a non-empty directory named `volume-delta-<last restored backup>.img` in the replica directory. This operation will lead to replica incremental restore failure later.
7 Create the 3rd backup for the regular volume
8. Wait for the DR volume incremental restore complete. And check:
    --> Check if the volume is `Degraded` state and the replica in step5 is ERROR.
    --> Check if volume.conditions[restore].status == True && volume.conditions[restore].reason == "RestoreInProgress".
9. See if the DR volume still works fine. (e.g. Trigger more incremental restore, trigger expansion)
10. Activate the DR volume and check the data.

**Comment by shuo-wu** [comment]
The PR for issue #1260 cannot completely solve this issue since there are some errors during the incremental restore not being recorded in the restore status. Then the longhorn manager cannot mark the related replica as failed.
The errors are not recorded in the sync agent server restore status:
1. Parameters check failure before restore status initialization.
2. Delta file check and cleanup error before starting the incremental restoration.
3. File coalesce error in the post inc restore function.
4. Replica reload error in the post inc restore function.
...
Besides, there is no related integration test covers those corner cases.

And the regular restore implementation has the similar issue, too.

If we want to cover those errors in the restore status, we need to check which error may be fatal for the replica (longhorn manager needs to fail the replica according to the restore status error msg) and which error won't affect the replica (longhorn manager doesn't need to fail the replica and simply retrying restore may be enough.). It means we need to refactor the whole restore, including incremental restore. I don't think we have time to do that in this release.

Actually I can fix the issue that triggered by the reproduce step, but I don't think that's what we want.

**Comment by yasker** [comment]
Have we fixed https://github.com/longhorn/longhorn/issues/1328 along with #1260 ?

Sounds like we need to move this out of 1.0.0.

**Comment by yasker** [comment]
@shuo-wu Can you assess the impact of missing the fix for this issue? E.g. in which case the user will have trouble, what's the workaround, etc.

**Comment by shuo-wu** [comment]
No, #1328 will be fixed along with other issues later.
The fix for #1260 would only handle parts of the restore failure cases. We will record all found cases in #1337

**Comment by yasker** [comment]
Code merged, need the pre-merge checklist.

**Comment by shuo-wu** [comment]
## Pre-merged Checklist
* [x] Does the PR include the explanation for the fix or the feature?
It's in the PR.

* [x] Is the backend code merged (Manager, Engine, Instance Manager, BackupStore, etc)?
The PR is at:
https://github.com/longhorn/backupstore/pull/43
https://github.com/longhorn/longhorn-engine/pull/501
https://github.com/longhorn/longhorn-manager/pull/609

* [x] Is the reproduce steps/test steps documented?
See the reproducing step.

* [x] Which areas/issues this PR might have potential impacts on?
Area: Manager & Engine
Issues: Restore failure handling

* ~~[x] **If the fix introduces the code for backward compatibility** Has a separate issue been filed with the label `release/obsolete-compatibility`?
The compatibility issue is filed at~~

* ~~[x] **If labeled: area/ui** Has the UI issue filed or ready to be merged?
The UI issue/PR is at~~

* ~~[x] **if labeled: require/doc** Has the necessary document PR submitted or merged?
The Doc issue/PR is at~~

* [x] **If labeled: require/automation-e2e** Has the end-to-end test plan been merged? Have QAs agreed on the automation test case?
The automation skeleton PR is at https://github.com/longhorn/longhorn-tests/pull/357
The automation test case PR is at https://github.com/longhorn/longhorn-tests/pull/357

* [x] **if labeled: require/automation-engine** Has the engine integration test been merged?
The engine automation PR is at https://github.com/longhorn/longhorn-engine/pull/501

* [x] ~~**if labeled: require/manual-test-plan** Has the manual test plan been documented?
The updated manual test plan is at~~

**Comment by khushboo-rancher** [comment]
Tried reproducing issue from https://github.com/longhorn/longhorn/issues/1188#issue-599512921 

On V1.0.0, The DR volume is stuck in ```Restore in progress``` forever. - It can't be activated.
Volume name - ```volume-restore-stuck-test-1```
[longhorn-support-bundle_e27c1f86-2285-474c-8127-9c41cc352773_2020-07-09T23-36-05Z.zip](https://github.com/longhorn/longhorn/files/4900021/longhorn-support-bundle_e27c1f86-2285-474c-8127-9c41cc352773_2020-07-09T23-36-05Z.zip)


On master ```97a869e```, DR volume gets degraded but healthy. Also, checked the data after activation, data is intact.
Volume name - ```volume-restore-stuck-test-2```
[longhorn-support-bundle_516cbf10-706f-4852-860b-729747916e3d_2020-07-09T23-36-20Z.zip](https://github.com/longhorn/longhorn/files/4900023/longhorn-support-bundle_516cbf10-706f-4852-860b-729747916e3d_2020-07-09T23-36-20Z.zip)

In above set up I don't see ```restoreStatus and lastRestoredBackup``` empty. 
@yasker Is this expected?

**Comment by shuo-wu** [comment]
1. The error log in the v1.0.0 test means the status stored in the engine process is inconsistent with the engine object status. Since the restore error cannot be handled correctly in v1.0.0, the whole volume will mess up after the test then many kinds of weird issues may get triggered. 
2. There will be a big refactor for the restore feature in the next release. This kind of error will be fixed after the reactor. 

In general, I don't think we need to worry about the error you triggered in v1.0.0 test now.

**Comment by khushboo-rancher** [comment]
Verified on v1.0.1-rc1

Validation- **Passed**

Followed the steps from https://github.com/longhorn/longhorn/issues/1188#issue-599512921 DR volume remains healthy. 
Verified the restored data.
